track/gps_client.py

```diff
@@ -0,0 +1,247 @@
+"""GPS location determination via gpsd.
+
+Defines a single class GPS that encapsulates the procedures necessary to obtain the current location
+as a client of the gpsd daemon.
+
+The code in this module works on the assumption that gpsd is already configured and running on the
+system.
+
+It is also highly suggested that ntpd is also already configured and running on the system, and
+preferably is configured to use the gps module's pps signal via gpsd as its time source. In any
+case, the validation code in this module will (optionally) enforce that the gps time matches the
+system time to a reasonable degree of precision.
+"""
+
+# http://manpages.ubuntu.com/manpages/bionic/man5/gpsd_json.5.html
+
+from collections import namedtuple
+from datetime    import datetime
+from enum        import Flag, IntEnum, auto
+from math        import inf, isinf, nan, isnan
+from time        import perf_counter
+#from types       import SimpleNamespace
+
+from astropy import units as u
+from astropy.coordinates import EarthLocation
+from astropy.time import Time as APTime
+
+import gps
+
+# improved enum-ized version of the 'MODE_' constants from the gps module
+class GPSFixType(IntEnum):
+    ZERO   = 0
+    NO_FIX = 1
+    FIX_2D = 2
+    FIX_3D = 3
+
+# TODO: use alternative type-hint syntax for these namedtuple declarations:
+#   from typing import NamedTuple
+#   class GPSValues(NamedTuple):
+#     lat: Angle # or whatever type we actually want to use for this
+#     lon: Angle # likewise
+#     # etc...
+GPSValues  = namedtuple('GPSValues',  ['lat', 'lon', 'alt', 'track', 'speed', 'climb', 'time'])
+GPSMargins = namedtuple('GPSMargins',                               ['speed', 'climb', 'time'])
+
+# TODO: write some nicer documentation for GPSMargins: (for all: inf means don't-care)
+# - speed: float: max valid deviation from zero for horizontal speed   (unit: meters/second)
+# - climb: float: max valid deviation from zero for vertical   speed   (unit: meters/second)
+# - time:  float: max valid deviation between gps time and system time (unit: seconds)
+
+# TODO: document the units on all the relevant fields here, and use types for them when appropriate
+#       (ask Brett for the best degree types and unit stuff to use here... astropy has good ones?)
```

Comment:
Yeah Astropy has good support for unit stuff. Of interest:

- `Quantity`: I think this is a base class from which a lot of other stuff is derived. From some reading I think these objects will keep track of units, so if you do some fancy expression like `10*u.m / 3*u.s` you'll get a `Quantity` object that stringifies to `<Quantity 3.3333333333333335 m / s>`.
- `Angle`: Stores angles. Can be initialized in degrees, radians, and probably other units. This is canonically done by e.g., `my_angle = Angle(13*u.deg)`
- Can do length with `u.m` for meters. I think by default an expression like `13*u.m` is a `Quantity` but I'm not certain of this.